palantir/gradle-graal
# Add Windows support

Hi there,

This PR adds Windows support to the only Gradle GraalVM plugin (this one!) - a much needed feature, I imagine (using Windows myself and experimenting with GraalVM). With this PR `gradle-graal` supports building native images and shared libraries on Windows, if you install the prerequisites GraalVM needs.

I haven't opened an issue, because discussing this code directly instead of theoretically seems better.

## How to test

Install the required dependencies for GraalVM to work on windows (see [substratevm/README.md](https://github.com/oracle/graal/blob/master/substratevm/README.md#quick-start)) and make sure GraalVM can successfully compile something standalone (remember to use the *Windows SDK Command Prompt* when running `native-image` directly).

- Clone [my `windows` branch](https://github.com/frozenice/gradle-graal/tree/windows)
- Run the Gradle task `publishToMavenLocal`
- Run the Gradle task `printVersion`, this will print something like `0.3.0-38-g2b56d74`
- In your project, add `mavenLocal()` and the locally published plugin to the buildscript dependencies, e.g.:
  ```gradle
  buildscript {
    repositories {
      mavenLocal()
      mavenCentral()
    }

    dependencies {
      classpath 'com.palantir.graal:gradle-graal:0.3.0-38-g2b56d74'
    }
  }
  ```
  (must be the first block in your build.gradle)
- Make sure you don't include `gradle-graal` via a `plugins` block or otherwise
- Configure the `graal` extension, e.g.:
  ```gradle
  graal {
    graalVersion "19.1.0"
    mainClass mainClassName
    outputName "graal-test"
  }
  ```
  (replace `mainClassName` with a string containing your entry class name if you haven't applied the `application` plugin)
- Run the Gradle task `nativeImage`, it will print the output location (the `sharedLibrary` task also works)

## Changes

### DownloadGraalTask

Enabling the download of the windows archive required adding a `WINDOWS` case to `getOperatingSystem` and a new `getArchiveExtension` method (it's `.zip` on Windows).

There is a new `[ext]` placeholder, which is used in `ARTIFACT_PATTERN_RELEASE_VERSION` (RC versions don't have a Windows variant) and `FILENAME_PATTERN`. It's replaced in `render`.

### ExtractGraalTask

The recent [version 19 support](https://github.com/palantir/gradle-graal/commit/b5567e451629ee75779b8f9b8c537e926e8f3350) added `getExecutable` and `getArchitectureSpecifiedBinaryPath`. I added Windows support to those (a bit of complexity arose, because some executables in the GraalVM Windows distribution have .cmd as an extension, others have .exe).

### BaseGraalCompileTask

I added an abstract method `getArchitectureSpecifiedOutputExtension` so subclasses (`NativeImageTask` and `SharedLibraryTask`) can specify the platform-dependant file extension for their output.

There is also a new `getArchitectureSpecifiedPathSeparator` method, because path separators differ (`;` in Windows, `:` on Unix).

A `WINDOWS` case was added to `getArchitectureSpecifiedBinaryPath`, so it finds the `native-image` command.

To overcome the need for running inside the *Windows SDK Command Prompt* (`SetEnv.cmd`) I replicated the environment variables configuration, stripped to a minimum (it sets the following environment variables when launching `native-image`: `CL`, `PATH`, `LIB`, `LIBPATH`, `INCLUDE`, `APPVER`).

To do this, it needs to read some values from the Windows registry. This is done in `readWindowsRegistryString`, which is a very simple (but sufficient) wrapper around reg.exe to query a value from the registry.

`fileSizeMegabytes` was lifted from `NativeImageTask` so it can be used in all subclasses.

### NativeImageTask

The overriden method `getArchitectureSpecifiedOutputExtension` returns `.exe` on Windows and an empty string on Linux and Mac.

`configurePlatformSpecifics` is called in the `@TaskAction`.

### SharedLibraryTask

Basically the same as `NativeImageTask`, the extensions in `getArchitectureSpecifiedOutputExtension` are different, though.

Oh, and I also added the `LogAction`, like in `NativeImageTask`.

### Tests

I upgraded some GraalVM version that's used in tests to the current `19.1.0` (it must be 19+ so the tests can run on Windows).

Some cross-platform enhancements include:
- Specifying `.exe` and `dll` for the output file extensions on Windows (could use `getArchitectureSpecifiedOutputExtension` too)
- Replacing hardcoded `\n`s with `System.lineSeparator()`
- Escape backslashes in the generated `gradle.properties` file
- ~Allowing both `localhost` and `127.0.0.1` in `requestUrl` returned by the mock server~
- Just compare the path of mock server requests (because the hostname is not always 127.0.0.1or even localhost).

The tests *allows specifying different RC graal version*, *test 1.0.0-rc5 nativeImage* and *can build shared libraries on 1.0.0-rc5* are now ignored on Windows, because there is no GraalVM RC version for Windows. 

The test *allows specifying different GA graal version* was split into a version for Windows and another version for Linux / Mac. There are some things that are different on Windows. This looks a bit like duplicate code (which it is), but parameterizing all the places looked a bit much.

## Notes

There are some baseline warnings. `SwitchStatementDefaultCase` was already present, `DefaultCharset` is new, but I'm not sure what to do about it. Of course it needs to use the default platform charset to read the command output! Using `Charset.defaultCharset()` produced another warning.

I haven't run tests on Linux or Mac yet.

Cheers 🍺 
David
